Implement a source adapter for raw data.

sebdesign opened the pull request:
It adapts a `TemporaryFile` source which encloses file contents as string, array or stream resource, and stores it in the temp directory until the end of the script

frasmage commented:
Hi @sebdesign,

So there are a few edge cases here which are not covered by the base package (content blob as a string and stream resource), though I'm not sure if I am sold of the implementation yet. I'll list out my thoughts:

- Why have two separate classes for `TemporaryFile` and `TemporaryFileAdapter`, why not just have the adapter handle the raw input and spit out the requisite fields? Your implementation of the adapter is currently just a passthrough. That said, there might be a argument for making a separate adapter for raw content and another for streams.
- Your implementation doubles up on disk writes as you copy the content to `/tmp` and then copy it into place. Granted, file uploads do the same, but that happens outside of the PHP process. In this case you already have the content separately, why do you need to write it twice?
- I'm fine with having a different methods than `fromSource()` to handle different cases (`fromStream()`, `fromRaw()`, etc.), though I fail to see the need for that second parameter. `SourceAdapters` provide the original filename as a convenient fallback if it is not otherwise specified. Given that the `useFilename()` method exists to manually apply a name, I don't see a reason force a title elsewhere. The source adapter could always return `"untitled"` as a default. With streams it is often possible to infer a filename with [stream_get_meta_data()](http://us.php.net/manual/en/function.stream-get-meta-data.php).

Let me know your thoughts!

sebdesign commented:
Thanks for your review!

The implementation needs some work, that's for sure.

- The reason for the `TemporaryFile` was to follow the paradigm from the other source adapters, which are mapped in the configuration. If the `TemporaryFileAdapter` can handle the raw data by itself without a source class, I'm not sure how it can be resolved through the `SourceAdapterFactory`. Maybe it doesn't have to, and we get rid of it from the configuration. That way the source has to be explicitly set through `fromContents()` and not through `fromSource()`.
Also a separate `StreamAdapter` and a `RawAdapter` sounds like a good idea.

- My first thought was to keep the raw contents in memory until they are stored by the `MediaUploader`. But I encountered the following issues:
    * The `SourceAdapterInterface` defines the `path` and `filename` methods which do not apply for raw file contents or streams (as far as I know).
    * How the `verifySource()` method is supposed  to check if the raw contents are valid? And if they aren't, it needs a path to construct the corresponding exception.
    * The `TemporaryFile` solves those issues since the file is present on the local filesystem, but I agree that's not ideal. Maybe `php://memory` or `php://temp` would solve this issue? 

- There is no filename obviously if there is no file stored, that's why the `TemporaryFile` requires it explicitly in order to respect the `SourceAdapterInterface`. The `SourceAdapterTest@test_it_adapts_filename` test expects a  file named `plank` at all times, but that's not possible because PHP generates random filenames when creating temp files. I agree that it would be best to have no filename at all until the user specifies it with `useFilename()` or return `untitled`, but that test would fail. Any thoughts on that?

frasmage commented:
> The reason for the `TemporaryFile` was to follow the paradigm from the other source adapters, which are mapped in the configuration. If the `TemporaryFileAdapter` can handle the raw data by itself without a source class, I'm not sure how it can be resolved through the `SourceAdapterFactory`. Maybe it doesn't have to, and we get rid of it from the configuration. That way the source has to be explicitly set through `fromContents()` and not through `fromSource()`.

The point of the `SourceAdapter` map in the config is that each item can be keyed to a different _pattern_ of input which behaves differently (e.g. different class hierarchy, different string regex). As far as I am aware, all blobs and all streams should behave the same way within the uploader, regardless of their content. So in these cases, there isn't really a pattern that can be applied to change the handling of the input. In effect, all blobs (or streams) are the same input as each other, where as not all strings should be handled the same way.

For that reason, perhaps we should just bind a single source adapter for each case to the service container (app) and have the associated `from*()` method of the uploader instantiate from there. This would still allow overrides (change container binding), but I doubt that would ever be necessary.

> The SourceAdapterInterface defines the `path` and `filename` methods which do not apply for raw file contents or streams (as far as I know).

For a blob, there is no path or filename (`null`?).

For a stream, both can be determined by the uri key of `stream_get_meta_data()`.

Filename is only used as a fallback, so it could be defaulted to something like `"untitled"` if no value is provided by the source or the user.

Path is currently only used in two places. The first use is in the exception message for a failed `verifySource()`, which shouldn't be an issue (see below). The second is in the `generateHash()` function. Perhaps this should be abstracted from the uploader to the `SourceAdapter`. For blobs, this should be pretty easy. IIRC, `mb_strlen($blob, '8bit');` should provide an accurate reading (this needs to be tested). I am not sure how to achieve this with streams, since we don't have the whole file until it has been copied to disk.

> How the `verifySource()` method is supposed to check if the raw contents are valid? And if they aren't, it needs a path to construct the corresponding exception.

`verifySource()` checks if the file exists and is readable. If we already have a blob of the file's entire contents, this would not be a concern and therefore always pass.

Correct me if I am wrong, but I am pretty sure `fopen()` would throw an exception if the source doesn't exist (local) or connect (remote), so I don't think you can get a resource handle for an invalid file. I think that this should also always pass as long as the stream isn't closed.

> The TemporaryFile solves those issues since the file is present on the local filesystem, but I agree that's not ideal. Maybe php://memory or php://temp would solve this issue?

TIL about those two streams. I think `php://memory` is potentially dangerous given the unknown size of the file. `php://temp` is potentially interesting, as it will handle small files in memory and use disk to tackle larger ones. Don't know enough about them yet to give a definitive answer.

> There is no filename obviously if there is no file stored, that's why the TemporaryFile requires it explicitly in order to respect the SourceAdapterInterface. The SourceAdapterTest@test_it_adapts_filename test expects a file named plank at all times, but that's not possible because PHP generates random filenames when creating temp files. I agree that it would be best to have no filename at all until the user specifies it with useFilename() or return untitled, but that test would fail. Any thoughts on that?

Those tests were written that way as a convenience, since the current set are all expected to have the same results regardless of the input. If we have a source that deviates from the norm, its tests would have to be handled manually.

sebdesign commented:
Hi @frasmage ,

I made some progress. I got rid of the temporary files, and extracted the functionality into a `StringAdapter` and a `StreamAdapter`. The `StringAdapter` works perfectly.

I'm stuck at the `StreamAdapter`, so I didn't include it in my last commit.

**File size**
In order to find the size there are two ways:
- we can use `fstat()` which doesn't work on remote files. See the note on the [docs](http://php.net/manual/en/function.fstat.php#refsect1-function.fstat-notes).
- we can attempt to read the resource contents via `stream_get_contents()`. This also doesn't work on for remote files, as the file has already been read. We can read read the contents with `fread` but it's slow and we can't rewind it back.
- check if the resource wrapper type is `http` and get the `Content-Length` from the wrapper_data. This actually works.

**Mime type**
Again, we need to read the file as a string in order to determine the mime type with `finfo->buffer()`. This is fine for local file streams, but not for remote files. In this case we have to check if the wrapper type is `http` and read the `Content-Type` from the header. This works as expected.

Maybe remote file stream resources shouldn't be allowed, since we have the `RemoteUrlAdapter` in place.

Also there are more resource types that could be useful for this package, like GD image resources, PDF documents, etc.

frasmage commented:
Hi @sebdesign,

StringAdapter looks good. One question: is `StringAdapter::getContents()` streaming to memory any more or less efficient than just returning the string value?

As for stream handling, given that we can read the wrapper type, perhaps we make more than one `StreamAdapter` mappable from the config. For example, make a `FileStreamAdapter` and `HttpStreamAdapter`, and have the SourceAdapterFactory assign as necessary. Each of those classes can then make use of the best available method for determining the various attributes.

This discussion is making me slightly reconsider how the string-based adapters are set up. I am wondering if some of them would be more efficient using streams.

sebdesign commented:
Hi @frasmage ,

My guess is opening a handle in `php://memory` is less efficient than returning the string value.
`Filesystem#put` is able to accept both, but `SourceAdapterTest#test_it_adapts_file_contents` would fail because it expects a `resource`. On the other hand `SourceAdapterInterface#contents` defines a `string` should be returned, but it's actually a resource. Is it ok to return a resource or a string depending on the adapter implementation? I think it breaks the strictness of the interface.

Indeed we could create adapters based on the stream wrapper type, or even refactor the existing adapters to open resources earlier in the process.

Going further, we could even register our own stream wrapper, e.g. `plank://mediable`, and use it  with our source along with all the metadata we can acquire in each case. I know guzzle does something similar, but it's more complex. I'm not sure it's a good idea though.

frasmage commented:
> My guess is opening a handle in php://memory is less efficient than returning the string value.
Filesystem#put is able to accept both, but SourceAdapterTest#test_it_adapts_file_contents would fail because it expects a resource. On the other hand SourceAdapterInterface#contents defines a string should be returned, but it's actually a resource. Is it ok to return a resource or a string depending on the adapter implementation? I think it breaks the strictness of the interface.

The interface was originally supposed to return a string when I wrote it, though I realized somewhere in the process that passing a stream is more memory efficient for larger files. I understand your concern, though given that the whole point of the adapter is to map it to the needs of the uploader, I don't mind too much with its outputs being as permissible as `Filesystem::put()`'s inputs. I'm on the fence about whether changing the interface to be stream only (`contents()` becomes a bit of a misnomer, maybe next major version rename to `open()` or `stream()`), or allowing multiple return types (bit of a code smell).

> Indeed we could create adapters based on the stream wrapper type, or even refactor the existing adapters to open resources earlier in the process.

I think this is definitely the way to go.

> Going further, we could even register our own stream wrapper, e.g. plank://mediable, and use it with our source along with all the metadata we can acquire in each case. I know guzzle does something similar, but it's more complex. I'm not sure it's a good idea though.

probably overkill for our purposes.

frasmage commented:
Also, the more I think about it, the more I want to break up `fromSource()`. At some point it may be worth doing:

```php
MediaUploader::fromUri('http://example.com');
MediaUploader::fromObject($request->file($foo));
MediaUploader::fromStream($handle);
MediaUploader::fromContents($blob);
```

have each type of input correspond to a different pattern map in the config.

We could probably leave `fromSource()` method and let it infer the type as well as possible, like it currently does. It would be easy to discern objects and streams, though URI/path and contents are all strings. Perhaps see if any URI patterns match, if not assume raw contents – though that might cause unexpected errors.

sebdesign commented:
Breaking up `fromSource()` looks like a good idea. But looking at the current architecture, it just delegates to the `SourceAdapterFactory` to adapt each source type. Having multiple `fromX` methods seems like the `MediaUploader` itself is doing the adaptation instead of the `SourceAdapterFactory`. So the question is, is it `SourceAdapterFactory`'s or `MediaUploader`'s responsibility/concern to choose the source adapter?

frasmage commented:
The factory still has a purpose of being the singleton instance which keeps track of the registered `SourceAdapter` patterns and classes. I would propose changing it such that the `MediaUploader` examines to type and `SourceAdapterFactory` examines to pattern, instead of the factory doing both. It wouldn't change the responsibility so much as add a level of indirection. 

I feel that this makes for an overall cleaner API, since the inputs the user provides can be stricter.

sebdesign commented:
Hi @frasmage ,

I've done some progress on the stream adapters.

The `SourceAdapterFactory` now accepts `resource` objects and adapts them depending on their wrapper implementation. I have included `file://`, `http://` and  `php://` wrapper adapters.

The I/O adapter is not bullet-proof, because all the streams cannot be used like we want. E.g. `php://output` is write-only. See the options at http://php.net/manual/en/wrappers.php.php.

Also, I was thinking that the `SourceAdapterFactory` should accept PSR-7 Streams, i.e. objects implementing the `Psr\Http\Message\StreamInterface`. This could be easily done with a class adapter. 

The problem is that we still need to know the stream wrapper type in order to implement the `SourceAdapterInteface` methods accordingly. The `*StreamAdapter` classes are already there, but the `SourceAdapterFactory` adapts `resource` objects instead of PSR-7 `StreamInterface` objects. It would be interesting if we could handle both.

As a note, the abstract `StreamAdapter` could easily be refactored to implement the PSR-7 `StreamInterface`, but I'm not sure if it can/should remain abstract. Maybe it should be a simple `Plank\Mediable\Stream` class implementing the `StreamInterface`.

When a `resource` object is given, the `SourceAdapterFactory` constructs a `Plank\Mediable\Stream` object. Then the `*StreamAdapter`s would adapt `StreamInterface` objects instead of raw `resource` objects, and that could be done for `StreamInterface` objects given by the user.

frasmage commented:
Hey @sebdesign, 

Looking very promising!

Initial thoughts:

-typo `getMedatata()`
- You still have reference to `TemporaryFile` in the docs.
- I wonder if StreamAdapter should throw an exception if the constructor input is not a resource, or even if it is not a stream.
- I think StreamAdapter should deny `isValid()` if the stream is not readable.
-I would not leave commented out classes in the config. Especially not ones that don't exists. There might be some value in leaving, say `// 'ftp://' => ...` to instruct on possible stream types, but populating a class name in our namespace insinuates that it already exists.
- I don't know enough about any of the other stream types to say, would they be particularly complex to adapt? If any are compatible at all we can make those a separate PR someday.

sebdesign commented on src/SourceAdapters/StreamAdapter.php:
```diff
@@ -0,0 +1,131 @@
+    {
+        if (is_resource($source) && get_resource_type($source) === 'stream') {
+            $this->resource = $source;
+            $this->source = new Stream($source);
```
In order to accept both resources and `StreamInterface` object, the resource is being wrapped in a `Plank\Mediable\Stream` object.

sebdesign commented on src/SourceAdapters/StreamAdapter.php:
```diff
@@ -0,0 +1,131 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function contents()
```
A similar check is done here because a `SourceInterface` return its contents as string, and there is no way to get the underlying resource.

sebdesign commented on tests/integration/SourceAdapters/SourceAdapterTest.php:
```diff
@@ -95,8 +117,12 @@ public function test_it_adapts_file_contents($adapter, $source)
-        $this->assertTrue(is_resource($contents));
-        $this->assertEquals(get_resource_type($contents), 'stream');
+
+        if (is_resource($contents)) {
```
Also this conditional seems like a code smell, because of the contents returned by the `StreamInterface` or the resource.

sebdesign commented:
Hi @frasmage ,

I've fixed the typos and updated the docs.
Also, the adapters have been refactored to accept PSR-7 `StreamInterface` objects. But there are some things I don't like, and I see some code smells. Please see the comments on the PR.

frasmage commented:
Hi @sebdesign,

Interesting, looks like there is a lot for me to sift through here. I will try to give this a thorough review in the next day or two when I have the chance.

frasmage commented:
Hi @sebdesign, 

Sorry for the delayed reply, been very busy lately.

I agree that conditionals in tests are a code smell. We should definitely do away with that. I think all three of your issues can be addressed by going the string route. Looking at the list of `SourceAdapters`:

- `FileAdapter`, `UploadedFileAdapter`, `RemoteUrlAdapter`, and `LocalFileAdapter` don't use stream anywhere except in `contents()`, just as easy return a string with `file_get_contents()`.
- `FileStreamAdapter`, `HttpStreamAdapter`, and `IoStreamAdapter` are already streams, but the resource handle is not directly accessible if input is a `StreamInterface`, could always return a string with `stream_get_contents()` or `StreamInterface::contents()`
- `StringAdapter` is already a string, goes through hoops to turn into a stream resource.

At this point I think I am I am more or less decided that returning a string across the board is the way to go. The fact that this is the approach taken by PSR7 is the nail in the coffin for me. I wouldn't worry about breaking backwards compatibility as 1. the interface already describes returning a string (current behaviour is out of sync), 2. these adapters are only really used as an internally, most users won't touch them directly and 3. any user that defines a custom source adapter can still return a resource as `$filesystem->put()` won't care.

Can we rename `StringAdapter` to `RawContentAdapter` or something like that, "string" is a very generic term, imo.

I see you more or less copied guzzlehttp/psr7/Stream, though I have a few reservations about that implementation. I don't love having to cast to string to get the full contents, would prefer a discrete method. Furthermore, `__toString()` is [not allowed to throw an exception](http://php.net/manual/en/language.oop5.magic.php#object.tostring).

One other thing that I was wondering is whether the various steps of our process could cause a non-seekable stream to fail. For example, could reading the content to infer the mime type, size, etc. cause `contents()` to not receive the full content? Whatever implementation we go with should cache the content output before returning it.

sebdesign commented:
Hi @frasmage ,

Thank you for taking the time to review my changes!
I agree with all your points. Regarding your last thought, you are right. Stream resources from remote locations do not support seeking. I ran a few tests and the file contents were empty.

We could still use `stream_get_contents` to cache the contents, but it's not possible for PSR-7 streams, because `__toString()` MUST rewind in order to get all the contents, which raises an exception that is caught and returns an empty string. On that note, using `fopen()` on remote files is not recommended for a number of reasons, including this issue.

This doesn't stop us to cache the contents of PSR-7 streams, because most of the time it won't be streams of remote files opened with `fopen()`; I believe all the implementations are using cURL under the hood, which is fine.

But when we are dealing with http stream `resource`s, we could call `stream_get_contents()` before transforming it into a `Stream` object to cache its contents.

Moving a step backwards, I'm not even sure we should support the http wrapper at all. Caching its contents in one place and reading its headers in another place seems awkward, plus it's not that efficient. I'd rather have a `CurlAdapter` for this case.

That should leave us with the `FileStreamAdapter` and the `IoStreamAdapter`, which share almost the same code. We could move their methods to the `StreamAdapter` parent class, and get rid of them, because all the low level operations can be handled by the `StreamInterface` implementations.

frasmage commented:
Could we not handle non-seekable streams with something like the following on the adapter? That way they read all of the content once, regardless of how seekable they are.

```php
protected function streamContents()
{
    if(is_null($this->contents){
        if ($this->stream->seekable()) {
            $this->contents = (string) $this->stream;
        } else {
            $this->contents = $this->stream->getContents();
        }
    }
    return $this->contents;
}
```

If remote streams need more thought, I am ok with separating them from the PR to get the rest live.

sebdesign commented:
The problem with the `else` clause is that it's not guaranteed that the whole file will be read. The whole file will be read if we're dealing with remote file streams, but for all the other cases (which we can't be aware), `getContents()` can throw an exception if the contents are not read, or if they are read, only the remaining bytes are returned.

frasmage commented:
Understood, though as long as the package's internal usage doesn't affect the content, I am fine with it. If a user provides invalid input, they should receive an exception. If they provide us a non-seekable stream that hasn't been started yet, there is no reason we shouldn't be able to accept it.

sebdesign commented:
Can I ask you a somehow related question?

The `adaptClass()` method on the `SourceAdapterFactory` gets the source's self and parent classes and compares them against the `class_adapters` keys. Why aren't you simply using the `instanceof` keyword instead?

```php
        foreach ($this->class_adapters as $class => $adapter) {
            if ($source instanceof $class) {
                return $adapter;
            }
        }
```

I'm asking because I'm trying trying to map the `Psr\Http\Message\StreamInterface` with the `StreamAdapter` and that doesn't work because it's an interface and not a concrete class.

frasmage commented:
@sebdesign, I don't recall why that was written that way. That should probably be instanceof.

sebdesign commented:
I've done some major refactoring on the wiring of the streams.

First, I figured out that the wrapper-specific stream adapter are not the best solution because: a) it's not a scalable solution because the need to support more wrappers would require to implement new adapters each time, b) the source adapters should be able to adapter their source without relying on their internals.

Following the SOLID paradigm, I've moved the duplication of each stream adapter to their parent `StreamAdapter` class, which is no longer abstract. Now it just asks its source about the information it needs, instead of diving into the stream internals. Furthermore, the internal logic is being moved closer to the data, which is the given `StreamInterface` class implementation. Both our `Stream` class and the user-provided objects are capable to handle the underlying stream, so we can safely honor the `StreamInterface` contract.

Regarding the http streams, instead of relying on the headers, which are not totally reliable, in the same way the headers of an `UploadedFile` are not safe, we are using the file contents to determine the mime type, the extension and the size.

Also, the `StreamAdapter` caches the contents of the stream, so we don't have to worry for unseekable streams. Using a `StreamResourceAdapter` for `resource`s on top of the `StreamAdapter` removes the conditionals on the parent class, and makes things more clear IMO.

frasmage commented:
Hi @sebdesign,

Looks ready for merging to me! However, travis is suddenly failing due to time out. I might need to refactor the tests to not run full migrations/filesystem purge on every test when they aren't necessarily used.

frasmage commented:
Hey @sebdesign,

I have refactored the tests, merged in your branch and released a new minor version.

Thank you very much for all of your hard work on this PR! I really appreciate the collaborative discussion that this sparked. I hope to see you around again in the future!

sebdesign commented:
Hi @frasmage ,

thank you very much for merging this! Your help was very valuable and insightful. I hope too we can collaborate in such meaningful manner in the future. Let me know if you have any other ideas or thoughts, and perhaps I can submit a PR.

Cheers!
